src/tests/helpers/page.ts

```diff
@@ -0,0 +1,241 @@
+import { JSHandle, Locator, Page } from "@playwright/test"
+
+type EventLog = [string, any, string | null]
+type MutationLog = [string, string | null, string | null]
+
+export function attributeForSelector(page: Page, selector: string, attributeName: string): Promise<string | null> {
+  return page.locator(selector).getAttribute(attributeName)
+}
+
+export function clickWithoutScrolling(page: Page, selector: string, options = {}) {
+  const element = page.locator(selector, options)
+
+  return element.evaluate((element) => element instanceof HTMLElement && element.click())
+}
+
+export function clearLocalStorage(page: Page): Promise<void> {
+  return page.evaluate(() => localStorage.clear())
+}
+
+export function disposeAll(...handles: JSHandle[]): Promise<void[]> {
+  return Promise.all(handles.map((handle) => handle.dispose()))
+}
+
+export function getFromLocalStorage(page: Page, key: string) {
+  return page.evaluate((storageKey: string) => localStorage.getItem(storageKey), key)
+}
+
+export function getSearchParam(url: string, key: string): string | null {
+  return searchParams(url).get(key)
+}
+
+export function hash(url: string): string {
+  const { hash } = new URL(url)
+
+  return hash
+}
+
+export async function hasSelector(page: Page, selector: string): Promise<boolean> {
+  return !!(await page.locator(selector).count())
+}
+
+export function innerHTMLForSelector(page: Page, selector: string): Promise<string> {
+  return page.locator(selector).innerHTML()
+}
+
+export async function isScrolledToSelector(page: Page, selector: string): Promise<boolean> {
+  const boundingBox = await page
+    .locator(selector)
+    .evaluate((element) => (element instanceof HTMLElement ? { x: element.offsetLeft, y: element.offsetTop } : null))
+
+  if (boundingBox) {
+    const { y: pageY } = await scrollPosition(page)
+    const { y: elementY } = boundingBox
+    const offset = pageY - elementY
+    return Math.abs(offset) < 2
+  } else {
+    return false
+  }
+}
+
+export function nextBeat() {
+  return sleep(100)
+}
+
+export function nextBody(_page: Page, timeout = 500) {
+  return sleep(timeout)
```

Comment:
@keithamus this is one of the bigger hurdles for the migration. The old version tracked the `<body>` element with instance state. Since playwright test blocks have no context, I'm not sure of the best way to synchronize on a change.